File: shared-lib/go/cards.go
//go:build !debug && !tests

package main

import (
	"C"
	"fmt"
	"time"
	"unsafe"

	"github.com/uhppoted/uhppote-core/types"
	"github.com/uhppoted/uhppote-core/uhppote"
)

func getCards(uu uhppote.IUHPPOTE, N *C.int, deviceID uint32) error {
	if N == nil {
		return fmt.Errorf("invalid argument (N) - expected valid pointer")
	}

	cards, err := uu.GetCards(deviceID)
	if err != nil {
		return err
	}

	*N = C.int(cards)

	return nil
}

func getCard(uu uhppote.IUHPPOTE, card *C.struct_Card, deviceID uint32, cardNumber uint32) error {
	if card == nil {
		return fmt.Errorf("invalid argument (card) - expected valid pointer")
	}

	c, err := uu.GetCardByID(deviceID, cardNumber)
	if err != nil {
		return err
	} else if c == nil {
		return fmt.Errorf("%v: no response to get-card %v", deviceID, cardNumber)
	}

	card.card_number = C.uint(c.CardNumber)
	card.from = C.CString(fmt.Sprintf("%v", c.From))
	card.to = C.CString(fmt.Sprintf("%v", c.To))

	doors := unsafe.Slice(card.doors, 4)

	doors[0] = C.uchar(c.Doors[1])
	doors[1] = C.uchar(c.Doors[2])
	doors[2] = C.uchar(c.Doors[3])
	doors[3] = C.uchar(c.Doors[4])

	return nil
}

func getCardByIndex(uu uhppote.IUHPPOTE, card *C.struct_Card, deviceID uint32, index uint32) error {
	if card == nil {
		return fmt.Errorf("invalid argument (card) - expected valid pointer")
	}

	response, err := uu.GetCardByIndex(deviceID, index)
	if err != nil {
		return err
	} else if response == nil {
		return fmt.Errorf("%v: no response to get-card-by-index %v", deviceID, index)
	}

	card.card_number = C.uint(response.CardNumber)
	card.from = C.CString(fmt.Sprintf("%v", response.From))
	card.to = C.CString(fmt.Sprintf("%v", response.To))

	doors := unsafe.Slice(card.doors, 4)

	doors[0] = C.uchar(response.Doors[1])
	doors[1] = C.uchar(response.Doors[2])
	doors[2] = C.uchar(response.Doors[3])
	doors[3] = C.uchar(response.Doors[4])

	return nil
}

func putCard(uu uhppote.IUHPPOTE, deviceID uint32, cardNumber uint32, from, to *C.char, doors *uint8) error {
	_from, err := time.Parse("2006-01-02", C.GoString(from))
	if err != nil {
		return fmt.Errorf("Invalid 'from' date (%v)", err)
	}

	_to, err := time.Parse("2006-01-02", C.GoString(to))
	if err != nil {
		return fmt.Errorf("Invalid 'to' date (%v)", err)
	}

	if doors == nil {
		return fmt.Errorf("invalid argument (doors) - expected valid pointer")
	}

	_doors := unsafe.Slice(doors, 4)

	card := types.Card{
		CardNumber: cardNumber,
		From:       (*types.Date)(&_from),
		To:         (*types.Date)(&_to),
		Doors: map[uint8]int{
			1: int(_doors[0]),
			2: int(_doors[1]),
			3: int(_doors[2]),
			4: int(_doors[3]),
		},
	}

	ok, err := uu.PutCard(deviceID, card)
	if err != nil {
		return err
	} else if !ok {
		return fmt.Errorf("%v: put-card failed", deviceID)
	}

	return nil
}

func deleteCard(uu uhppote.IUHPPOTE, deviceID uint32, cardNumber uint32) error {
	deleted, err := uu.DeleteCard(deviceID, cardNumber)
	if err != nil {
		return err
	} else if !deleted {
		return fmt.Errorf("%v: delete-card %v failed", deviceID, cardNumber)
	}

	return nil
}

func deleteCards(uu uhppote.IUHPPOTE, deviceID uint32) error {
	deleted, err := uu.DeleteCards(deviceID)
	if err != nil {
		return err
	} else if !deleted {
		return fmt.Errorf("%v: delete-cards failed", deviceID)
	}

	return nil
}

File: shared-lib/tests/c/cards.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tests.h"
#include "uhppoted.h"

bool getCards() {
    const char *tag = "get-cards";
    int N;

    if (get_cards(DEVICE_ID, &N) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {
        {
            .field = "card count",
            .type = "uint32",
            .value.uint32.expected = 39,
            .value.uint32.value = N,
        },
    };

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool getCard() {
    const char *tag = "get-card";
    card card;

    if (get_card(DEVICE_ID, CARD_NUMBER, &card) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    // clang-format off
    const result resultset[] = {
        { .field = "card number",      .type = "uint32", .value.uint32.expected = 8165538,      .value.uint32.value = card.card_number },
        { .field = "card 'from' date", .type = "string", .value.string.expected = "2022-01-01", .value.string.value = card.from },
        { .field = "card 'to' date",   .type = "string", .value.string.expected = "2022-12-31", .value.string.value = card.to },
        { .field = "card doors[1]",    .type = "uint8",  .value.uint8.expected = 0,             .value.uint8.value = card.doors[0] },
        { .field = "card doors[2]",    .type = "uint8",  .value.uint8.expected = 1,             .value.uint8.value = card.doors[1] },
        { .field = "card doors[3]",    .type = "uint8",  .value.uint8.expected = 31,            .value.uint8.value = card.doors[2] },
        { .field = "card doors[4]",    .type = "uint8",  .value.uint8.expected = 75,            .value.uint8.value = card.doors[3] },
    };
    // clang-format on

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool getCardByIndex() {
    const char *tag = "get-card-by-index";
    card card;

    if (get_card_by_index(DEVICE_ID, CARD_INDEX, &card) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    // clang-format off
    const result resultset[] = {
        { .field = "card number",      .type = "uint32", .value.uint32.expected = 8165538,      .value.uint32.value = card.card_number },
        { .field = "card 'from' date", .type = "string", .value.string.expected = "2022-01-01", .value.string.value = card.from },
        { .field = "card 'to' date",   .type = "string", .value.string.expected = "2022-12-31", .value.string.value = card.to },
        { .field = "card doors[1]",    .type = "uint8",  .value.uint8.expected = 0,             .value.uint8.value = card.doors[0] },
        { .field = "card doors[2]",    .type = "uint8",  .value.uint8.expected = 1,             .value.uint8.value = card.doors[1] },
        { .field = "card doors[3]",    .type = "uint8",  .value.uint8.expected = 31,            .value.uint8.value = card.doors[2] },
        { .field = "card doors[4]",    .type = "uint8",  .value.uint8.expected = 75,            .value.uint8.value = card.doors[3] },
    };
    // clang-format on

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool putCard() {
    const char *tag = "put-card";
    uint8_t doors[4] = {0, 1, 31, 75};

    if (put_card(DEVICE_ID, CARD_NUMBER, "2022-01-01", "2022-12-31", (uint8_t *)doors) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {};

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool deleteCard() {
    const char *tag = "delete-card";

    if (delete_card(DEVICE_ID, CARD_NUMBER) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {};

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool deleteCards() {
    const char *tag = "delete-cards";

    if (delete_cards(DEVICE_ID) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {};

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}
